crates/nu-protocol/src/engine/pwd_per_drive.rs

```diff
@@ -0,0 +1,289 @@
+use crate::engine::{EngineState, Stack};
+use std::path::{Path, PathBuf};
+
+// For file system command usage
+/// Proxy/Wrapper for
+/// nu_path::expand_path_with<P, Q>(path, relative_to, expand_tilde);
+///
+/// Usually if a command opens one file or directory, it uses
+/// nu_path::expand_path_with::<P, Q>(p, r, t) to expand '~','.' etc.; replacing it with
+/// nu_protocol::engine::fs_client::expand_path_with(stack, engine_state, p, r t) will
+/// first check if the path is relative for a drive;
+/// Commands that accept multiple files/directories as parameters usually depend on Glob,
+/// after near future revised Glob collection implementation done, all file system commands
+/// will support PWD-per-drive.
+pub fn expand_path_with<P, Q>(
+    _stack: &Stack,
+    _engine_state: &EngineState,
+    path: P,
+    relative_to: Q,
+    expand_tilde: bool,
+) -> PathBuf
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    #[cfg(windows)]
+    if let Some(abs_path) = windows::expand_pwd(_stack, _engine_state, path.as_ref()) {
+        return abs_path;
+    }
+
+    nu_path::expand_path_with::<P, Q>(path, relative_to, expand_tilde)
+}
+
+#[cfg(windows)]
+pub mod windows {
+    use super::*;
+    use crate::{FromValue, Value};
+
+    pub trait EnvMaintainer {
+        fn maintain(&mut self, key: String, value: Value);
+    }
+
+    macro_rules! impl_env_maintainer {
```

Comment:
Can you remove this macro and just expand it. We don't really like macros if we don't have to have them.

Reply:
OK, just copy same impl for EngineState and Stack

Reply:
yup, that's how i see it.